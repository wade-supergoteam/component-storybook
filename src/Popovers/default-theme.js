// @flow

import { sassRgba } from "../utils/sassRgba";
import { black, white } from "../utils/config/colors";

export const zIndexPopover = "1060";

export const popoverInnerPadding = "1px";
export const popoverBg = white;
export const popoverMaxWidth = "276px";
export const popoverBorderWidth = "1px";
export const popoverBorderColor = sassRgba(black, 0.2);
export const popoverBoxShadow = `0 5px 10px ${sassRgba(black, 0.2)}`;

export const popoverTitleBbg = "#f7f7f7";
export const popoverTitlePaddingX = "14px";
export const popoverTitlePaddingY = "8px";
export const popoverTitleBorderBottom = "1px solid #ebebeb";

export const popoverContentPaddingX = "14px";
export const popoverContentPaddingY = "9px";

export const popoverArrowWidth = "10px";
export const popoverArrowColor = white;

export const popoverArrowOuterWidth = `11px`;
export const popoverArrowOuterColor = sassRgba(popoverBorderColor, 0.25);

export const fontSizeSm = ".875rem";
export const borderRadius = ".25rem";
