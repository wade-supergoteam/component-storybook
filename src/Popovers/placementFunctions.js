import { css } from "styled-components";

import themeProp from "../utils/theme";
import { popoverArrowOuterWidth, popoverArrowWidth, popoverArrowColor, popoverBorderColor } from "./default-theme";

const placementTop = offsetLeft => css`
  div:first-child {
    left: ${() => (offsetLeft ? `${offsetLeft}px` : "50%")};
    border-width: ${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    border-bottom-width: 0;
    border-top-color: ${themeProp("popoverBorderColor", popoverBorderColor)};
    bottom: 0;
    margin-bottom: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    margin-left: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};

    &::after {
      content: "";
      border-width: ${themeProp("popoverArrowWidth", popoverArrowWidth)};
      top: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
      margin-left: -${themeProp("popoverArrowWidth", popoverArrowWidth)};
      border-top-color: ${themeProp("popoverArrowColor", popoverArrowColor)};
    }
  }
`;

const placementBottom = offsetLeft => css`
  div:first-child {
    left: ${() => (offsetLeft ? `${offsetLeft}px` : "50%")};
    border-width: ${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    border-top-width: 0;
    border-bottom-color: ${themeProp("popoverBorderColor", popoverBorderColor)};
    top: 0;
    margin-top: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    margin-left: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};

    &::after {
      content: "";
      border-width: ${themeProp("popoverArrowWidth", popoverArrowWidth)};
      bottom: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
      margin-left: -${themeProp("popoverArrowWidth", popoverArrowWidth)};
      border-bottom-color: ${themeProp("popoverArrowColor", popoverArrowColor)};
    }
  }
`;

const placementRight = offsetTop => css`
  div:first-child {
    top: ${() => (offsetTop ? `${offsetTop}px` : "50%")};
    border-width: ${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    border-left-width: 0;
    border-right-color: ${themeProp("popoverBorderColor", popoverBorderColor)};
    left: 0;
    margin-left: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    margin-top: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};

    &::after {
      content: "";
      top: ${() => (offsetTop ? `${offsetTop}px` : "50%")};
      border-width: ${themeProp("popoverArrowWidth", popoverArrowWidth)};
      border-right-color: ${themeProp("popoverArrowColor", popoverArrowColor)};
      margin-top: -${themeProp("popoverArrowWidth", popoverArrowWidth)};
      margin-left: calc(-${popoverArrowWidth} + 1px);
    }
  }
`;

const placementLeft = offsetTop => css`
  div:first-child {
    top: ${() => (offsetTop ? `${offsetTop}px` : "50%")};
    border-width: ${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    border-right-width: 0;
    border-left-color: ${themeProp("popoverBorderColor", popoverBorderColor)};
    right: 0;
    margin-right: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    margin-top: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};

    &::after {
      content: "";
      top: ${() => (offsetTop ? `${offsetTop}px` : "50%")};
      border-width: ${themeProp("popoverArrowWidth", popoverArrowWidth)};
      bottom: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
      margin-top: -${themeProp("popoverArrowWidth", popoverArrowWidth)};
      border-left-color: ${themeProp("popoverArrowColor", popoverArrowColor)};
      margin-left: -${themeProp("popoverArrowOuterWidth", popoverArrowOuterWidth)};
    }
  }
`;

export { placementTop, placementBottom, placementRight, placementLeft };
