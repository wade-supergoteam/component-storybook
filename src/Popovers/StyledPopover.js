import styled from "styled-components";

import themeProp from "../utils/theme";
import * as defaultTheme from "./default-theme";

import { placementTop, placementBottom, placementRight, placementLeft } from "./placementFunctions";

const {
  popoverMaxWidth,
  zInderxPopover,
  popoverInnerPadding,
  popoverBg,
  popoverBorderWidth,
  popoverBorderColor,
  popoverBoxShadow,
  popoverTitlePaddingX,
  popoverTitlePaddingY,
  popoverTitleBbg,
  popoverTitleBorderBottom,
  fontSizeSm,
  borderRadius,
} = defaultTheme;

const popoverPlacement = ({ placement, arrowOffset }) => {
  switch (placement) {
    case "top":
      return placementTop(arrowOffset);
    case "bottom":
      return placementBottom(arrowOffset);
    case "right":
      return placementRight(arrowOffset);
    case "left":
      return placementLeft(arrowOffset);
    default:
      return null;
  }
};

const StyledPopover = styled.div`
  position: absolute;
  top: ${({ positionTop }) => `${positionTop}px` || 0};
  left: ${({ positionLeft }) => `${positionLeft}px` || 0};
  z-index: ${themeProp("zInderxPopover", zInderxPopover)};
  display: block;
  max-width: ${themeProp("popoverMaxWidth", popoverMaxWidth)};
  padding: ${themeProp("popoverInnerPadding", popoverInnerPadding)};
  font-size: ${themeProp("fontSizeSm", fontSizeSm)};
  word-wrap: break-word;
  background-color: ${themeProp("popoverBg", popoverBg)};
  background-clip: padding-box;
  border: ${themeProp("popoverBorderWidth", popoverBorderWidth)} solid ${themeProp("popoverBorderColor", popoverBorderColor)};
  border-radius: ${themeProp("borderRadius", borderRadius)};
  box-shadow: ${themeProp("popoverBoxShadow", popoverBoxShadow)};

  div:first-child,
  div:first-child:after {
    position: absolute;
    display: block;
    width: 0;
    height: 0;
    border-color: transparent;
    border-style: solid;
  }
  ${props => popoverPlacement(props)};
  h3 {
    margin: 0;
    padding: ${themeProp("popoverTitlePaddingY", popoverTitlePaddingY)} ${themeProp("popoverTitlePaddingX", popoverTitlePaddingX)};
    font-size: 14px;
    background-color: ${themeProp("popoverTitleBbg", popoverTitleBbg)};
    border-bottom: ${themeProp("popoverTitleBorderBottom", popoverTitleBorderBottom)};
    border-radius: 5px 5px 0 0;
  }
  div:last-child {
    padding: 9px 14px;
  }
`;

export default StyledPopover;
