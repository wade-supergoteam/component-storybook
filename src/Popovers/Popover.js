import React from "react";
import PropTypes from "prop-types";

import StyledPopover from "./StyledPopover";
import Heading from "../Heading";

const Popover = ({ title, children, ...props }) => (
  <StyledPopover {...props}>
    {title && <Heading as="h3">{title}</Heading>}
    <div>{children}</div>
  </StyledPopover>
);

Popover.defaultProps = {
  title: "",
};

Popover.propTypes = {
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
};

export default Popover;
