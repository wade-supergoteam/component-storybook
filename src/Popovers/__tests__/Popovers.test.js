import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import Popovers from "../Popover";
import StyledPopover from "../StyledPopover";

test("Default Popovers", () => {
  const wrapper = renderer.create(
    <Popovers title="Top" placement="top" arrowOffset="30" positionLeft="0">
      Child goes here
    </Popovers>,
  );
  expect(wrapper).toMatchSnapshot();
});

test("Default Popovers bottom", () => {
  const wrapper = renderer.create(
    <Popovers title="bottom" placement="bottom" arrowOffset="30" positionLeft="20">
      Child goes here
    </Popovers>,
  );
  expect(wrapper).toMatchSnapshot();
});

test("Default Popovers left", () => {
  const wrapper = renderer.create(
    <Popovers title="left" placement="left" arrowOffset="30" positionLeft="0">
      Child goes here
    </Popovers>,
  );
  expect(wrapper).toMatchSnapshot();
});

test("Default Popovers left 50%", () => {
  const wrapper = renderer.create(
    <Popovers title="left" placement="left" arrowOffset="30" positionLeft="50">
      Child goes here
    </Popovers>,
  );
  expect(wrapper).toMatchSnapshot();
});

test("Default Popovers right", () => {
  const wrapper = renderer.create(
    <Popovers title="right" placement="right" arrowOffset="30" positionLeft="20">
      Child goes here
    </Popovers>,
  );
  expect(wrapper).toMatchSnapshot();
});

test("Default Popovers right", () => {
  const wrapper = renderer.create(<Popovers title="right">Child goes here</Popovers>);
  expect(wrapper).toMatchSnapshot();
});

const customTheme = {
  popoverTitleBbg: "green",
  popoverBg: "#cce4ff",
  popoverBorderColor: "green",
};

test("Customized Popovers", () => {
  const wrapper = renderer.create(<ThemeProvider theme={customTheme} />);
  expect(wrapper).toMatchSnapshot();
});

test("Styled Popovers", () => {
  const wrapper = renderer.create(<StyledPopover />);
  expect(wrapper).toMatchSnapshot();
});

test("placementRight Popovers", () => {
  const wrapper = renderer.create(<placementRight />);
  expect(wrapper).toMatchSnapshot();
});

test("placementLeft Popovers", () => {
  const wrapper = renderer.create(<placementLeft />);
  expect(wrapper).toMatchSnapshot();
});
