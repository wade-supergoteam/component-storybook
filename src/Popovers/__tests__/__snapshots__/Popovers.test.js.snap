// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customized Popovers 1`] = `null`;

exports[`Default Popovers 1`] = `
<div
  className="sc-bdVaJa eQdIS"
>
  <h3
    className="sc-bwzfXH gRgMOy"
  >
    Top
  </h3>
  <div>
    Child goes here
  </div>
</div>
`;

exports[`Default Popovers bottom 1`] = `
<div
  className="sc-bdVaJa eIxeJC"
>
  <h3
    className="sc-bwzfXH gRgMOy"
  >
    bottom
  </h3>
  <div>
    Child goes here
  </div>
</div>
`;

exports[`Default Popovers left 1`] = `
<div
  className="sc-bdVaJa iGiOLk"
>
  <h3
    className="sc-bwzfXH gRgMOy"
  >
    left
  </h3>
  <div>
    Child goes here
  </div>
</div>
`;

exports[`Default Popovers left 50% 1`] = `
<div
  className="sc-bdVaJa ZTHBH"
>
  <h3
    className="sc-bwzfXH gRgMOy"
  >
    left
  </h3>
  <div>
    Child goes here
  </div>
</div>
`;

exports[`Default Popovers right 1`] = `
<div
  className="sc-bdVaJa detaLK"
>
  <h3
    className="sc-bwzfXH gRgMOy"
  >
    right
  </h3>
  <div>
    Child goes here
  </div>
</div>
`;

exports[`Default Popovers right 2`] = `
<div
  className="sc-bdVaJa byMprg"
>
  <h3
    className="sc-bwzfXH gRgMOy"
  >
    right
  </h3>
  <div>
    Child goes here
  </div>
</div>
`;

exports[`Styled Popovers 1`] = `
<div
  className="sc-bdVaJa byMprg"
/>
`;

exports[`placementLeft Popovers 1`] = `<placementLeft />`;

exports[`placementRight Popovers 1`] = `<placementRight />`;
